### PR TITLE
BUG: do not crash on a callable grouper returning tuples

### LINKED ISSUE #22257: get_group(...) fails for groupby(...) based on a function
#### Code Sample, a copy-pastable example if possible

```python
"""
IMPORTANT: The issue happens with pandas==0.21.1 but does NOT happen with pandas==0.19
"""
# Define simple dataframe, from Pandas doc ;)
# http://pandas.pydata.org/pandas-docs/stable/groupby.html#groupby-sorting
df3 = pd.DataFrame({'X' : ['A', 'B', 'A', 'B'], 'Y' : [1, 4, 3, 2]})

# Define a simple function for grouping
def func(df, index):
    return (df['X'].loc[index], df['Y'].loc[index] % 2)

# Make sure the groupby worked, and check the group names
for name, df in df3.groupby(by=lambda index: func(df3, index), sort=False):
    print(name)
    print(df)

# Execute the code that triggers an error
df3.groupby(by=lambda index: func(df3, index), sort=False).get_group(('A', 1))
```
#### Problem description
The code raises: `NotImplementedError: isna is not defined for MultiIndex`
**Note**: That issue happens with pandas==0.21.1. It does NOT happen with pandas==0.19.

#### Expected Output
df3.groupby(...).get_group(('A', 1)) should not raise an error but return the group ('A', 1):
```
('A', 1)
   X  Y
0  A  1
2  A  3
```
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-30-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: fr_FR.UTF-8
LOCALE: fr_FR.UTF-8

pandas: 0.21.1
pytest: None
pip: 9.0.1
setuptools: 39.0.1
Cython: None
numpy: 1.13.3
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #22257
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Notice I didn't have time to test which commit caused the regression.

### CODE REVIEW

File: pandas/tests/groupby/test_grouping.py [hunk 1]
```diff
@@ -202,6 +202,20 @@ def test_grouper_creation_bug(self):
...
+        # GH 22257
+        df = pd.DataFrame({'X': ['A', 'B', 'A', 'B'], 'Y': [1, 4, 3, 2]})
+
+        def func(index):
+            return (df['X'].loc[index], df['Y'].loc[index] % 2)
+
+        gb = df.groupby(by=func, sort=False)
+
+        name, expected = list(gb)[0]
+        result = gb.get_group(name)
```
Comment (WillAyd) [inline pandas/tests/groupby/test_grouping.py]: Can you be more explicit about what should be returned here? I think retrieving the element from the GroupBy object and subsequently passing that to `get_group` could potentially hide bugs in the future

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -508,6 +508,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
             warnings.warn(msg, FutureWarning, stacklevel=5)
             key = list(key)
 
+    if callable(key) or isinstance(key, dict):
```
Comment (WillAyd) [inline pandas/core/groupby/grouper.py]: Is the `isinstance(key, dict)` new here? If so probably want a test case to cover

File: pandas/tests/groupby/test_grouping.py [hunk 3]
```diff
@@ -346,7 +360,7 @@ def test_groupby_grouper_f_sanity_checked(self):
         # when the elements are Timestamp.
         # the result is Index[0:6], very confusing.
 
-        pytest.raises(AssertionError, ts.groupby, lambda key: key[0:6])
+        pytest.raises(ValueError, ts.groupby, lambda key: key[0:6])
```
Comment (WillAyd) [inline pandas/tests/groupby/test_grouping.py]: What is the reason for this change?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_grouping.py (hunk 3 above)
Comment (toobaz) [reply pandas/tests/groupby/test_grouping.py]: The single callable/mapping case used to follow the same code path as lists of callables/mappings, while it now follows the same code path as e.g. lists of ``Series``. The two happen to check the length of the grouper in different ways (the former should probably be changed to ``ValueError`` too, but we might decide to remove it altogether in #22278).

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (toobaz) [reply pandas/core/groupby/grouper.py]: (done - expanded the test)

File: pandas/tests/groupby/test_grouping.py (hunk 3 above)
Comment (toobaz) [reply pandas/tests/groupby/test_grouping.py]: Notice this test will also change when we remove the absurd ``DatetimeIndex.map`` implementation (see #22312 ) but that is presumably not happening in a bugfix release.

File: doc/source/whatsnew/v0.23.5.txt [hunk 4]
```diff
@@ -20,10 +20,11 @@ and bug fixes. We recommend that all users upgrade to this version.
 Fixed Regressions
 ~~~~~~~~~~~~~~~~~
 
-- Constructing a DataFrame with an index argument that wasn't already an
+- Constructing a :class:`DataFrame` with an index argument that wasn't already an
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.5.txt]: let' move to 0.24

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (jreback) [reply pandas/core/groupby/grouper.py]: can you integrate this with the below (e.g. i/elif maybe enough)?
can add a top-levevel comment on what is going on here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
closing as stale, but pls reopen if can continue
